Sources/SourceKitLSP/Swift/MacroExpansion.swift

```diff
@@ -0,0 +1,122 @@
+  var edits: [MacroExpansionEdit]
+
+  init(title: String, uri: DocumentURI, refactoringEdits: [RefactoringEdit]) {
+    self.title = title
+    self.uri = uri
+    self.edits = refactoringEdits.compactMap { refactoringEdit in
+      if refactoringEdit.bufferName.isEmpty {
+        logger.error("Unable to retrieve some parts of the expansion")
+        return nil
+      }
+
+      return refactoringEdit
+    }
+  }
+}
+
+extension SwiftLanguageService {
+  /// Handles the `ExpandMacroCommand`.
+  ///
+  /// Makes request to sourcekitd and wraps the result into a `MacroExpansion`
+  /// and then makes `ShowDocumentRequest` to the client side for each
+  /// expansion to be displayed.
+  ///
+  /// - Parameters:
+  ///   - expandMacroCommand: The `ExpandMacroCommand` that triggered this
+  /// request.
+  ///
+  /// - Returns:
+  ///   - an `[MacroExpansionEdit]` with the necessary edits and buffer name as
+  /// a `LSPAny`
+  func expandMacro(
+    _ expandMacroCommand: ExpandMacroCommand
+  ) async throws -> LSPAny {
+    guard let sourceKitLSPServer else {
+      // `SourceKitLSPServer` has been destructed. We are tearing down the
+      // language server. Nothing left to do.
+      throw ResponseError.unknown("Connection to the editor closed")
+    }
+
+    guard let sourceFileURL = expandMacroCommand.textDocument.uri.fileURL else {
+      throw ResponseError.unknown("Given URL is not a file URL")
+    }
+
+    let expansion = try await self.refactoring(expandMacroCommand)
+
+    for macroEdit in expansion.edits {
+      let macroExpansionBufferDirectoryName = macroEdit.bufferName.dropLast(6)  // buffer name without ".swift"
+
+      let macroExpansionBufferDirectoryURL = self.generatedMacroExpansionsPath.appendingPathComponent(
+        String(macroExpansionBufferDirectoryName)
+      )
+      do {
+        try FileManager.default.createDirectory(at: macroExpansionBufferDirectoryURL, withIntermediateDirectories: true)
+      } catch {
+        throw ResponseError.unknown(
+          "Failed to create directory for Macro Expansion Buffer at path: \(macroExpansionBufferDirectoryURL.path)"
+        )
+      }
+
+      let macroExpansionFileName = sourceFileURL.deletingPathExtension().lastPathComponent  // name of the source file
+      let macroExpansionPositionRangeIndicator =
+        "L\(macroEdit.range.lowerBound.line)C\(macroEdit.range.lowerBound.utf16index)-L\(macroEdit.range.upperBound.line)C\(macroEdit.range.upperBound.utf16index)"  // github permalink notation for position range
+
+      let macroExpansionFilePath = macroExpansionBufferDirectoryURL.appendingPathComponent(
+        "\(macroExpansionFileName)_\(macroExpansionPositionRangeIndicator).\(sourceFileURL.pathExtension)"
+      )
+
+      let macroExpansionDocURI = DocumentURI(macroExpansionFilePath)
+
+      do {
+        try macroEdit.newText.write(to: macroExpansionFilePath, atomically: true, encoding: String.Encoding.utf8)
```

Comment:
You can drop the `String.Encoding`.

```suggestion
        try macroEdit.newText.write(to: macroExpansionFilePath, atomically: true, encoding: .utf8)
```